tools/mage/dev.go

```diff
@@ -70,10 +76,90 @@ func execDockerCompose(args ...string) error {
 	return err
 }
 
+func execDockerComposeWithOutput(filepath string, args ...string) error {
+	output, err := sh.Output("docker-compose", dockerComposeFlags(args...)...)
+	if err != nil {
+		return err
+	}
+	message := []byte(output)
+	err = ioutil.WriteFile(filepath, message, 0644)
+	return err
+}
+
+// SQLStart starts the SQL database of the development environment.
+func (Dev) SQLStart() error {
+	if mg.Verbose() {
+		fmt.Println("Starting SQL database")
+	}
+	if err := execDockerCompose(append([]string{"up", "-d"}, sqlDatabase)...); err != nil {
+		return err
+	}
+	return execDockerCompose("ps")
+}
+
+// SQLStop stops the SQL database of the development environment.
+func (Dev) SQLStop() error {
+	if mg.Verbose() {
+		fmt.Println("Stopping SQL database")
+	}
+	return execDockerCompose(append([]string{"stop"}, sqlDatabase)...)
+}
+
+// SQLMakeSnapshot stores the current cockroach data folder for later restores.
+func (Dev) SQLMakeSnapshot() error {
+	if mg.Verbose() {
+		fmt.Println("Making DB snapshot")
+	}
+	to := filepath.Join(devDataDir, "cockroach-snap")
+	from := filepath.Join(devDataDir, "cockroach")
+	if err := os.RemoveAll(to); err != nil {
+		return err
+	}
+	return sh.Copy(from, to)
+}
+
+// SQLRestoreSnapshot restores the previously taken snapshot, thus restoring all previously
+// snapshoted databases.
+func (d Dev) SQLRestoreSnapshot() error {
+	mg.Deps(Dev.SQLStop)
+	if mg.Verbose() {
+		fmt.Println("Restoring DB snapshot")
+	}
+	to := filepath.Join(devDataDir, "cockroach")
+	from := filepath.Join(devDataDir, "cockroach-snap")
+	if err := os.RemoveAll(to); err != nil {
+		return err
+	}
+	if err := sh.Copy(from, to); err != nil {
+		return err
+	}
+	return d.SQLStart()
+}
+
+// DBDump performs an SQL database dump of the dev database to the .cache folder.
+func (Dev) DBDump() error {
+	if mg.Verbose() {
+		fmt.Println("Saving database dump")
+	}
+	if err := os.MkdirAll(".cache", 0755); err != nil {
+		return err
+	}
+	return execDockerComposeWithOutput(filepath.Join(".cache", "sqldump.sql"), "exec", "-T", "cockroach", "./cockroach", "dump", devDatabaseName, "--insecure")
```

Comment:
Not a big fan of this, as it assumes we always use CockroachDB (which we don't; in CI we use PostgreSQL)

Reply:
But this actually works in CI 🤷‍♂️

Reply:
Split line for readability - see e.g. https://github.com/TheThingsNetwork/lorawan-stack/blob/31119a0db8e9a7933e021195623f50ba5420f2df/tools/mage/dev.go#L135-L140